File: Matlabox/stats/modzscore.m
function modz = modzscore(obsData, sampleData, dim)
    %calcualte modified z-score of obsData to sampleData (distribution)
    %
    % if sampleData is a matrix, dim suggests which dimension is "variable"
    % and which is samples of the variable
    % if obsData is a vector or a matrix, and sampleData is a matrix, the
    % "variable" dimension of obsData much match sampleData:
    %   obsData: variable x obs, sampleData: variable x samples, modz:
    %       variable x obs
    %   obsData: 1 x obs, sampleData: variable x sample, modz: 1 x obs
    %   
    %
    
    medAbsDev = mad(sampleData, 1, dim) * 1.486; %median abs dev
    meanAbsDev = mad(sampleData, 0, dim) * 1.253314;
    i = medAbsDev == 0;
    medAbsDev(i) = meanAbsDev(i);
    
    med = nanmedian(sampleData, dim);
    
    tmp = bsxfun(@minus, obsData, med);
    modz = bsxfun(@rdivide, tmp, medAbsDev);
    
    if any(medAbsDev == 0)
        infidx = bsxfun(@and, medAbsDev==0, tmp > 0);
        modz(infidx) = Inf;
        infidx = bsxfun(@and, medAbsDev==0, tmp < 0);
        modz(infidx) = -Inf;
    end
    

File: rscripts/modzscore.r
modzscore = function(observedData, cohortData, MARGIN) {
    # R's mad already applies 1.486 factor
    medAbsDev = apply(cohortData, MARGIN, FUN= function(x) mad(x, na.rm= TRUE))
    meanAbsDev = apply(cohortData, MARGIN, FUN= function(x) mad(x, center= mean(x), constant= 1.253314, na.rm= TRUE))
    idx = medAbsDev == 0
    medAbsDev[idx] = meanAbsDev[idx]

    m = apply(cohortData, MARGIN, FUN= function(x) median(x, na.rm= TRUE))
    tmp = sweep(observedData, MARGIN, m, FUN= '-')
    modz = sweep(tmp, MARGIN, medAbsDev, FUN= '/')

    if (any(medAbsDev == 0) ) {
        idx = sweep(tmp > 0, MARGIN, medAbsDev == 0, FUN= '&')
        modz[idx] = Inf
        idx = sweep(tmp < 0, MARGIN, medAbsDev == 0, FUN= '&')
        modz[idx] = -Inf
    }
    return (modz)
}
